### ISSUE TITLE
Terraform Modules In Azure DevOps

### ISSUE THREAD

**Issue by psyvision** [tinglesoftware/dependabot-azure-devops#520]
We have been trying to get this extension working with private Terraform modules hosted in Azure DevOps git and having problems accessing the module repository.

We are running using the `:latest` docker image tag.  We used to have the problem with `.git` being appended to repository URLs but that is now resolved.

We have two repositories both hosted under `{org}`:
- Project
- Infrastructure

We have tried both a `git` and a `terraform-registry` registry type as I'm not exactly sure which it would be, this is not a Terraform registry server, but both get the same error.

.github/dependabot.yml:
```yaml
version: 2
updates:
  - package-ecosystem: "terraform"
    directory: "/deploy/infrastructure"
    schedule:
      interval: "daily"
    open-pull-requests-limit: 10
    registries:
      - git-infra
      #- terra-infra

registries:
  git-infra:
    type: git
    url: https://dev.azure.com
    username: x-access-token
    password: PAT:${{AZDO_PAT}}

  terra-infra:
    type: terraform-registry
    url: https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure
    token: PAT:${{AZDO_PAT}}
```

We include the repository as a source in our pipeline such that the access token gets permission to access the index.  Note this works for our other pipelines that use Terraform `plan`/`apply`:

pipeline.yml:
```yml
trigger: none

resources:
  repositories:
    - repository: infrastructure
      type: git
      name: Infrastructure/Infrastructure

stages:
  - stage: Dependabot
    displayName: Dependabot Stage
    condition: always()
    jobs:
      - job: Dependabot
        displayName: Dependabot
        condition: succeededOrFailed()
        continueOnError: true
        pool:
          name: "Azure Pipelines"
          vmImage: "ubuntu-latest"

        steps:
          - task: dependabot@1
            displayName: Dependabot
            inputs:
              useConfigFile: true
              openPullRequestsLimit: 10
            env:
              AZDO_PAT: $(System.AccessToken)
```

The output log from the run is:

```
/usr/bin/docker run --rm -i -e DEPENDABOT_PACKAGE_MANAGER=terraform -e DEPENDABOT_OPEN_PULL_REQUESTS_LIMIT=10 -e DEPENDABOT_DIRECTORY=/deploy/infrastructure -e DEPENDABOT_EXTRA_CREDENTIALS=[{"type":"git","url":"[https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure","registry":null,"username":"x-access-token","password":"PAT:***"},{"type":"terraform_registry","url":"https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure","registry":null,"token":"PAT:***"}](https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure%22,%22registry%22:null,%22username%22:%22x-access-token%22,%22password%22:%22PAT:***%22%7D,%7B%22type%22:%22terraform_registry%22,%22url%22:%22https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure%22,%22registry%22:null,%22token%22:%22PAT:***%22%7D)] -e DEPENDABOT_FAIL_ON_EXCEPTION=true -e AZURE_ORGANIZATION={org} -e AZURE_PROJECT=Project -e AZURE_REPOSITORY=Project -e AZURE_ACCESS_TOKEN=*** -e AZURE_MERGE_STRATEGY=squash tingle/dependabot-azure-devops:latest
....
<snip>
....
fatal: not a git repository: /home/dependabot/dependabot-core/../.git/modules/dependabot-core
fatal: not a git repository: /home/dependabot/dependabot-core/../.git/modules/dependabot-core
Using '[https://dev.azure.com:443/'](https://dev.azure.com/') as API endpoint
Working in {org}/Project/_git/Project, 'default' branch under '/deploy/infrastructure' directory
Looking for configuration file in the repository ...
🌍 --> GET https://dev.azure.com/{org}/Project/_apis/git/repositories/Project
🌍 <-- 200 https://dev.azure.com/{org}/Project/_apis/git/repositories/Project
🌍 --> GET https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/stats/branches?name=master
🌍 <-- 200 https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/stats/branches?name=master
🌍 --> GET https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/items?path=.github/dependabot.yml&versionDescriptor.versionType=commit&versionDescriptor.version=708242262ce1a066d1749ccddef235b8759d3c97
🌍 <-- 200 https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/items?path=.github/dependabot.yml&versionDescriptor.versionType=commit&versionDescriptor.version=708242262ce1a066d1749ccddef235b8759d3c97
Using configuration file at '/.github/dependabot.yml' 😎
Cloning repository into /home/dependabot/dependabot-updater/tmp/{org}/Project/_git/Project
Found 9 dependency file(s) at commit 708242262ce1a066d1749ccddef235b8759d3c97
 - /deploy/infrastructure/webapp.tf
 - /deploy/infrastructure/db.tf
 - /deploy/infrastructure/variables.tf
 - /deploy/infrastructure/locals.tf
 - /deploy/infrastructure/main.tf
 - /deploy/infrastructure/backend.tf
 - /deploy/infrastructure/outputs.tf
 - /deploy/infrastructure/versions.tf
 - /deploy/infrastructure/.terraform.lock.hcl
Parsing dependencies information
Found 5 dependencies
 - db::azure::{org}/Infrastructure/_git/Infrastructure ()
 - db_server::azure::{org}/Infrastructure/_git/Infrastructure ()
 - hashicorp/azurerm (3.14.0)
 - keyvault::azure::{org}/Infrastructure/_git/Infrastructure ()
 - web_app::azure::{org}/Infrastructure/_git/Infrastructure ()
🌍 --> GET https://dev.azure.com/{org}/_apis/connectionData
🌍 <-- 200 https://dev.azure.com/{org}/_apis/connectionData
🌍 --> GET https://dev.azure.com/{org}/Project/_apis/git/repositories/Project
🌍 <-- 200 https://dev.azure.com/{org}/Project/_apis/git/repositories/Project
🌍 --> GET https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/pullrequests?api-version=6.0&searchCriteria.status=active&searchCriteria.creatorId=93a56af8-3932-4083-b5cb-b70b50108830&searchCriteria.targetRefName=refs/heads/master
🌍 <-- 200 https://dev.azure.com/{org}/Project/_apis/git/repositories/Project/pullrequests?api-version=6.0&searchCriteria.status=active&searchCriteria.creatorId=93a56af8-3932-4083-b5cb-b70b50108830&searchCriteria.targetRefName=refs/heads/master
Checking if db::azure::{org}/Infrastructure/_git/Infrastructure  needs updating
🌍 --> GET https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure/info/refs?service=git-upload-pack
🌍 <-- 404 https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure/info/refs?service=git-upload-pack
/home/dependabot/dependabot-core/common/lib/dependabot/git_metadata_fetcher.rb:78:in `fetch_upload_pack_for': The following git URLs could not be retrieved: https://{org}@dev.azure.com/{org}/Infrastructure/_git/Infrastructure (Dependabot::GitDependenciesNotReachable)
	from /home/dependabot/dependabot-core/common/lib/dependabot/git_metadata_fetcher.rb:17:in `upload_pack'
	from /home/dependabot/dependabot-core/common/lib/dependabot/git_metadata_fetcher.rb:51:in `head_commit_for_ref'
	from /home/dependabot/dependabot-core/common/lib/dependabot/git_commit_checker.rb:99:in `head_commit_for_local_branch'
	from /home/dependabot/dependabot-core/common/lib/dependabot/git_commit_checker.rb:92:in `head_commit_for_current_branch'
	from /home/dependabot/dependabot-core/terraform/lib/dependabot/terraform/update_checker.rb:118:in `latest_version_for_git_dependency'
	from /home/dependabot/dependabot-core/terraform/lib/dependabot/terraform/update_checker.rb:17:in `latest_version'
	from /home/dependabot/dependabot-core/terraform/lib/dependabot/terraform/update_checker.rb:37:in `updated_requirements'
	from /home/dependabot/dependabot-core/common/lib/dependabot/update_checkers/base.rb:292:in `changed_requirements'
	from /home/dependabot/dependabot-core/common/lib/dependabot/update_checkers/base.rb:278:in `requirements_up_to_date?'
	from /home/dependabot/dependabot-core/common/lib/dependabot/update_checkers/base.rb:35:in `up_to_date?'
	from bin/update-script.rb:570:in `block in <main>'
	from bin/update-script.rb:533:in `each'
	from bin/update-script.rb:533:in `<main>'
##[error]The process '/usr/bin/docker' failed with exit code 1
Finishing: Dependabot
```

Any ideas on the 404?  It to me suggests an authentication issue but I'm not quite sure where the problem lies

**Comment by mburumaxwell** [comment]
It is likely that this was fixed in #506 specifically for `terraform-registry`. A [new release is now available](https://github.com/tinglesoftware/dependabot-azure-devops/releases/tag/0.15.0), you can try and report back.

**Comment by psyvision** [comment]
I updated the tag to 0.15 and ran a pipeline with what looks to be the same error:

```
Unable to find image 'ghcr.io/tinglesoftware/dependabot-updater:0.15' locally
0.15: Pulling from tinglesoftware/dependabot-updater
...
<snip>
...
Checking if db::azure::{org}/Infrastructure/_git/Infrastructure  needs updating
🌍 --> GET https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure/info/refs?service=git-upload-pack
🌍 <-- 404 https://dev.azure.com/{org}/Infrastructure/_git/Infrastructure/info/refs?service=git-upload-pack
/home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/git_metadata_fetcher.rb:78:in `fetch_upload_pack_for': The following git URLs could not be retrieved: https://{org}@dev.azure.com/{org}/Infrastructure/_git/Infrastructure (Dependabot::GitDependenciesNotReachable)
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/git_metadata_fetcher.rb:17:in `upload_pack'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/git_metadata_fetcher.rb:51:in `head_commit_for_ref'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/git_commit_checker.rb:99:in `head_commit_for_local_branch'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/git_commit_checker.rb:92:in `head_commit_for_current_branch'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/terraform/lib/dependabot/terraform/update_checker.rb:118:in `latest_version_for_git_dependency'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/terraform/lib/dependabot/terraform/update_checker.rb:17:in `latest_version'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/terraform/lib/dependabot/terraform/update_checker.rb:37:in `updated_requirements'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/update_checkers/base.rb:292:in `changed_requirements'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/update_checkers/base.rb:278:in `requirements_up_to_date?'
	from /home/dependabot/dependabot-updater/vendor/ruby/3.1.0/bundler/gems/dependabot-core-0672e3b867bb/common/lib/dependabot/update_checkers/base.rb:35:in `up_to_date?'
	from bin/update-script.rb:570:in `block in <main>'
	from bin/update-script.rb:533:in `each'
	from bin/update-script.rb:533:in `<main>'
##[error]The process '/usr/bin/docker' failed with exit code 1
Finishing: Dependabot
```

**Comment by mburumaxwell** [comment]
Kindly provide a public repro in Azure DevOps

**Comment by psyvision** [comment]
Created an Infra project - https://dev.azure.com/psyvision/Infra/_git/Infra
Created a project which uses it - https://dev.azure.com/psyvision/Project/_git/Project

Log below - works OK, which I expected as the repo is publicly accessible - I'm going to make the Infra private to see if it replicates the issue.

```
Using '[https://dev.azure.com:443/'](https://dev.azure.com/') as API endpoint
Working in psyvision/Project/_git/Project, 'default' branch under '/' directory
Looking for configuration file in the repository ...
🌍 --> GET https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project
🌍 <-- 200 https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project
🌍 --> GET https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/stats/branches?name=master
🌍 <-- 200 https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/stats/branches?name=master
🌍 --> GET https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/items?path=.github/dependabot.yml&versionDescriptor.versionType=commit&versionDescriptor.version=7e33783e89f9a5b2dd496b0ff07d0fad9437ed3e
🌍 <-- 200 https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/items?path=.github/dependabot.yml&versionDescriptor.versionType=commit&versionDescriptor.version=7e33783e89f9a5b2dd496b0ff07d0fad9437ed3e
Using configuration file at '/.github/dependabot.yml' 😎
Cloning repository into /home/dependabot/dependabot-updater/tmp/psyvision/Project/_git/Project
Found 2 dependency file(s) at commit 7e33783e89f9a5b2dd496b0ff07d0fad9437ed3e
 - /main.tf
 - /.terraform.lock.hcl
Parsing dependencies information
Found 2 dependencies
 - group::azure::psyvision/_git/Infra ()
 - hashicorp/azurerm (3.45.0)
🌍 --> GET https://dev.azure.com/psyvision/_apis/connectionData
🌍 <-- 200 https://dev.azure.com/psyvision/_apis/connectionData
🌍 --> GET https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project
🌍 <-- 200 https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project
🌍 --> GET https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/pullrequests?api-version=6.0&searchCriteria.status=active&searchCriteria.creatorId=d91a8d3d-dd09-44e9-a3e8-d6e6c21b322d&searchCriteria.targetRefName=refs/heads/master
🌍 <-- 200 https://dev.azure.com/psyvision/Project/_apis/git/repositories/Project/pullrequests?api-version=6.0&searchCriteria.status=active&searchCriteria.creatorId=d91a8d3d-dd09-44e9-a3e8-d6e6c21b322d&searchCriteria.targetRefName=refs/heads/master
Checking if group::azure::psyvision/_git/Infra  needs updating
🌍 --> GET https://dev.azure.com/psyvision/_git/Infra/info/refs?service=git-upload-pack
🌍 <-- 200 https://dev.azure.com/psyvision/_git/Infra/info/refs?service=git-upload-pack
No update needed for group::azure::psyvision/_git/Infra 
Checking if hashicorp/azurerm 3.45.0 needs updating
🌍 --> GET https://registry.terraform.io/.well-known/terraform.json
🌍 <-- 200 https://registry.terraform.io/.well-known/terraform.json
🌍 --> GET https://registry.terraform.io/v1/providers/hashicorp/azurerm/versions
🌍 <-- 200 https://registry.terraform.io/v1/providers/hashicorp/azurerm/versions
No update needed for hashicorp/azurerm 3.45.0
Done
Finishing: Dependabot
```

**Comment by psyvision** [comment]
Made the Infra repo private and the build with the error can be seen here:

https://dev.azure.com/psyvision/Project/_build/results?buildId=6&view=logs&j=fc40e090-047c-59dc-b11b-bf41948db735&t=6aeda54f-00c7-5efd-2577-0106410dc434

If you want an invite to see the private project or know any org/project level config etc then let me know.

**Comment by mburumaxwell** [comment]
@psyvision could this be a permissions issue cross-repository or cross-project?

Could you expose the Infra repo for me to clone and repro?

**Comment by psyvision** [comment]
@mburumaxwell I have made it public again.

We use the same project setup (albeit maybe some slightly different permissions private under the actual organisation) to run our actual terraform plan/apply so we know it _should/does_ work.  In that project pipeline, we specify the repository in the same manner such that the `System.AccessToken` contains the correct claims (?) to be able to access the Infra repo.  The theory here with dependabot is that the same PAT passed through (see original config) should enabled the Infra clone to happen within the docker container.

I know there are a few debug environment vars (is it `EXCON` something?) that could be set, is it possible to use that to provide more info?  I guess it may not expose what auth header (if any) is actually being used.

**Comment by mburumaxwell** [comment]
@psyvision 
Is the git reference correct?
https://dev.azure.com/psyvision/Project/_git/Project?path=/main.tf&version=GBmaster&line=11&lineEnd=12&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents

You have `git::https://dev.azure.com/psyvision/_git/Infra//resource_group` which is missing the project name and also file paths are specified using a query parameter in Azure DevOps.

I am not too conversant with Terraform but the first issue I worked on for it is #347 and it seems similar to your setup.
The `source` sed there is `git::https://dev.azure.com/tingle/dependabot/_git/repro-347-tf-modules?path=/src/rg&ref=0.1.0` and I think yours should be similar to that. This is especially the case because the error in your pipeline run points to a missing source or sorts.

I suggest you work with `git::https://dev.azure.com/psyvision/Infra/_git/Infra?path=/resource_group&ref=master`

When you change the URL you will also need to generate the correct lock file. The registries in the configuration file will then not be needed because your access token will be used by default.

I managed to get this to work:
- Repo 1: https://dev.azure.com/tingle/dependabot/_git/repro-520-Infra
- Repo 2: https://dev.azure.com/tingle/dependabot/_git/repro-520-Project
- Run: https://dev.azure.com/tingle/dependabot/_build/results?buildId=57490&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=1f93c1a3-255b-505a-53c9-c382114aabf5

Let me know if this helps.

**Comment by psyvision** [comment]
Hi @mburumaxwell - I meant to reply to this as I have finally got it working.  I don't fully understand how, but Microsoft corrected a bug in Azure DevOps pipelines which means you now **must** specify which jobs use which resources (repositories) via the `uses` aspect of the YAML.  Now that I have added this in to my Dependabot pipeline all is working as expected.  This post explains it a bit more - https://developercommunity.visualstudio.com/t/System-Access-Token-not-being-given-repo/10323773#T-ND10328571 but doesn't fully allude to how or why this worked before and what they've actually now changed that "broke" it.
